src/typecheck/destructuring.rs

```diff
@@ -26,23 +37,35 @@ enum TypecheckMode {
 }
 
 /// Build a `UnifType` from a `Destruct` pattern. The type of each "leaf"
-/// identifier will be assigned based on the `leaf_type` argument. The
-/// current possibilities are for each leaf to have type `Dyn` or to be
-/// assigned a fresh unification variable.
+/// identifier will be assigned based on the `mode` argument. The
+/// current possibilities are for each leaf to have type `Dyn`, to use an
+/// explicit type annotation, or to be assigned a fresh unification variable.
 fn build_pattern_type(
     state: &mut State,
+    ctxt: &Context,
     pat: &RecordPattern,
-    leaf_type: TypecheckMode,
-) -> UnifRecordRows {
-    fn new_leaf_type(state: &mut State, leaf_type: TypecheckMode) -> UnifType {
-        match leaf_type {
+    mode: TypecheckMode,
+) -> Result<UnifRecordRows, UnifError> {
+    fn new_leaf_type(
+        state: &mut State,
+        ctxt: &Context,
+        mode: TypecheckMode,
+        ty_annot: Option<LabeledType>,
+    ) -> UnifType {
+        match mode {
             TypecheckMode::Walk => mk_uniftype::dynamic(),
```

Comment:
Something I just realised is that I'm not entirely sure how we should handle the case where `ty_annot` is `Some` but we're in `Walk` mode... presumably the presence of a type annotation in the pattern should actually move us into typechecking mode? :thinking: 

I guess one way to do this would be to match on `ty_annot` first, and then only check the `mode` if it's `None`.

Happy to hear other opinions on this though.

Reply:
> In my opinion type annotations don't make much sense in a pattern (at least in walk mode)

I think I can see a way they _could_ make sense, which would be to include the annotated type as the type of the field in the record type returned from `build_pattern_type`. That then gets unified with the inferred type of the value being destructured, and a type error would be raised if they differed. 

That said, I have no opinions on what the _utility_ of this behaviour would be, if any.

I think I agree that the current state of this PR, where annotations in walk mode are ignored during typechecking & then converted to contracts for evaluation, is probably a sensible approach.

Reply:
> I think I can see a way they could make sense, which would be to include the annotated type as the type of the field in the record type returned from build_pattern_type. That then gets unified with the inferred type of the field in the value being destructured, and a type error would be raised if they differed.

But in _walk mode_, there's no unification happening, right?

In the meantime, I've convinced myself that the distinction does obviously make sense in a typed block (and your implementation makes a dinstinction between a type annotation and a contract annotation). The difference between a contract annotation and a type annotation is usually that type annotations turns typechecking on (they are equivalent as far as evaluation is concerned), but as I said above, I don't see a natural semantics for such a thing when the type annotation is inside the pattern (there might be one! But right now, I don't see it).

Reply:
Ah, yes, you're totally right about not unifying in walk mode. 

I guess in that case it makes sense just to leave this as-is for now.

Reply:
> I guess in that case it makes sense just to leave this as-is for now.

Agreed!